server/aspects/rules/scala/scala_info.bzl

```diff
@@ -48,31 +52,62 @@ def extract_scalatest_classpath(rule_attr):
         extract_from_attr("_scalatest_reporter")
     )
 
-def extract_scala_info(target, ctx, output_groups, **kwargs):
+def extract_scala_info_mainline(target, ctx, output_groups):
+    scala_info = {}
+    rule_attr = ctx.rule.attr
+    common_scalac_opts = ctx.toolchains[MAINLINE_SCALA_TOOLCHAIN].scalacopts
+    if hasattr(rule_attr, "_scalac"):
+        scalac = rule_attr._scalac
+        compiler_classpath = find_scalac_classpath(scalac.default_runfiles.files.to_list())
+        if compiler_classpath:
+            scala_info["compiler_classpath"] = map(file_location, compiler_classpath)
+            if is_external(scalac):
+                update_sync_output_groups(output_groups, "bsp-sync-artifact", depset(compiler_classpath))
+    scala_info["scalac_opts"] = common_scalac_options + getattr(ctx.rule.attr, "scalacopts", [])
+    scala_info["scalatest_classpath"] = extract_scalatest_classpath(ctx.rule.attr)
+    return scala_info
+    
+def extract_scala_info_annex(target, ctx, output_groups):
+    scala_info = {}
+    rule_attr = ctx.rule.attr
+    scala_configuration = ctx.toolchains[SCALA_TOOLCHAIN].scala_configuration
+    common_scalac_options = scala_configuration.global_scalacopts
+
+    classpath_files = []
+    for target in scala_configuration.compiler_classpath:
+        for file in target[JavaInfo].runtime_output_jars:
```

Comment:
wont it fail for bazel 8 repos? you need to explicitly import the provider. You can use `get_java_info` from `server/aspects/rules/java/java_info.bzl.template` it has all the needed magic